cli/src/main/kotlin/utils/AuthenticationUtils.kt

```diff
@@ -0,0 +1,84 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.eclipse.apoapsis.ortserver.cli.utils
+
+import com.github.ajalt.clikt.core.ProgramResult
+
+import io.ktor.client.plugins.auth.Auth
+import io.ktor.client.plugins.auth.providers.BearerTokens
+import io.ktor.client.plugins.auth.providers.bearer
+import io.ktor.client.plugins.defaultRequest
+
+import org.eclipse.apoapsis.ortserver.cli.COMMAND_NAME
+import org.eclipse.apoapsis.ortserver.cli.model.AuthenticationStorage
+import org.eclipse.apoapsis.ortserver.cli.model.HostAuthenticationDetails
+import org.eclipse.apoapsis.ortserver.client.OrtServerClient
+import org.eclipse.apoapsis.ortserver.client.OrtServerClient.Companion.JSON
+import org.eclipse.apoapsis.ortserver.client.auth.AuthService
+import org.eclipse.apoapsis.ortserver.client.createDefaultHttpClient
+import org.eclipse.apoapsis.ortserver.client.createOrtHttpClient
+
+/**
+ * Create an authenticated ORT Server client using the stored authentication. Returns null if the user is not
+ * authenticated.
+ */
+fun createOrtServerClient() = AuthenticationStorage.get()?.let { createOrtServerClient(it) }
+
+/**
+ * Create a new instance of the ORT server client using the given [authDetails] and configure an HTTP client
+ * with the necessary authentication.
+ */
+private fun createOrtServerClient(authDetails: HostAuthenticationDetails): OrtServerClient {
+    val client = createOrtHttpClient(JSON) {
+        defaultRequest {
+            url(authDetails.baseUrl)
+        }
+
+        install(Auth) {
+            bearer {
+                loadTokens {
+                    BearerTokens(
+                        authDetails.tokens.access,
+                        authDetails.tokens.refresh
+                    )
+                }
+
+                refreshTokens {
+                    val updatedTokenInfo = runCatching {
+                        val auth = AuthService(
+                            client = createDefaultHttpClient(JSON),
+                            tokenUrl = authDetails.tokenUrl,
+                            clientId = authDetails.clientId
+                        )
+
+                        auth.refreshToken(authDetails.tokens.refresh)
+                    }.getOrElse {
+                        println("Failed to refresh authentication. Please run '$COMMAND_NAME auth login'.")
```

Comment:
It feels wrong to me to do console output in this utility function (and ideally this should use `echo()` anyway, but it lacks the terminal context to do so).

How about instead using `getOrThrow()` here and leaving it the callers to do console output / map to the desired type of exception?

Reply:
> It feels wrong to me to do console output in this utility function

Yes, I agree.
However, the error message is printed in the `refreshToken` lambda.
That means the caller (the CLI part), would need to wrap every _usage_ of the `OrtServerClient` into a try catch, in order to handle this in the same way, and not only at the time of creation of the client.

Alternatively, the CLI could offer another abstraction layer that handles these errors.

I'm not sure if this is worth the effort, or am I missing a simpler solution?

Reply:
> That means the caller (the CLI part), would need to wrap every usage of the OrtServerClient into a try catch, in order to handle this in the same way, and not only at the time of creation of the client.

Could this not be done in the `main` function?